### ISSUE TITLE
Memory allocation problem while training

### ISSUE THREAD

**Issue by thanasio** [Microsoft/CNTK#625]
Hi,

i am trying to train a dataset from a 27 GB file of 115 features x 29,097,256 lines on a workstation with a geforce GTX 750 ti gpu and 16 GB of RAM.

Here is my simple .cntk configuration file:

RootDir = ".."

ConfigDir = "$RootDir$/Config"
DataDir   = "$RootDir$/Data"
OutputDir = "$RootDir$/Output"
ModelDir  = "$OutputDir$/Models"

deviceId = 0

command = Simple_Demo_Train

precision = "float"
traceLevel = 0
modelPath = "$ModelDir$/simple.dnn"
outputNodeNames = ScaledLogLikelihood
numFeatures = 115
output = 4

#######################################
#  TRAINING CONFIG                    #
#######################################

Simple_Demo_Train = [
    action = "train"

    SimpleNetworkBuilder = [
        layerSizes = $numFeatures$:80*3:$output$
        trainingCriterion = "CrossEntropyWithSoftmax"
        evalCriterion = "ErrorPrediction"
        layerTypes = "Tanh"
        initValueScale = 1.0
        applyMeanVarNorm = true
        uniformInit = true
        needPrior = true
    ]

    SGD = [
        epochSize = 0 
        minibatchSize = 128*2	
        learningRatesPerMB = 0.05*2
	autoAdjustMiniBatch = true
	autoAdjustLR=SearchBeforeEpoch	
		
        momentumPerMB = 0.9
        dropoutRate = 0.0
        maxEpochs = 300
    ]
       

    reader = [
        readerType = "UCIFastReader"
	customDelimiter=","
        file = "$DataDir$/features_train.csv"
        miniBatchMode = "Full"
        verbosity = 1  

        features = [
            dim = $numFeatures$       
            start = 1    
        ]

        labels = [
            start = 0    
            dim = 1      
            labelDim = $output$  
            labelMappingFile = "$DataDir$/labelsmap.txt"
        ]
    ]
]

The problem i get at the end is:
=====================================================================
configparameters: infusSimpleNet.cntk:traceLevel=0
<<<<<<<<<<<<<<<<<<<< PROCESSED CONFIG WITH ALL VARIABLES RESOLVED <<<<<<<<<<<<<<<<<<<<
Commands: Simple_Demo_Train
Precision = "float"
CNTKModelPath: ../Output/Models/simple.dnn
CNTKCommandTrainInfo: Simple_Demo_Train : 300
CNTKCommandTrainInfo: CNTKNoMoreCommands_Total : 300
##############################################################################
#                                                                            #
# Action "train"                                                             #
#                                                                            #
##############################################################################
CNTKCommandTrainBegin: Simple_Demo_Train
SimpleNetworkBuilder Using GPU 0
Creating virgin network.
Microsoft::MSR::CNTK::GPUMatrix<ElemType>::SetUniformRandomValue (GPU): creating curand object with seed 1, sizeof(ElemType)==4
Post-processing network...
7 roots:
    CrossEntropyWithSoftmax = CrossEntropyWithSoftmax()
    EvalErrorPrediction = ErrorPrediction()
    InvStdOfFeatures = InvStdDev()
    MeanOfFeatures = Mean()
    PosteriorProb = Softmax()
    Prior = Mean()
    ScaledLogLikelihood = Minus()
Validating network. 30 nodes to process in pass 1.
Validating --> labels = InputValue() :  -> [4 x *]
Validating --> W3 = LearnableParameter() :  -> [4 x 80]
Validating --> W2 = LearnableParameter() :  -> [80 x 80]
Validating --> W1 = LearnableParameter() :  -> [80 x 80]
Validating --> W0 = LearnableParameter() :  -> [80 x 115]
Validating --> features = InputValue() :  -> [115 x *]
Validating --> MeanOfFeatures = Mean (features) : [115 x *] -> [115]
Validating --> InvStdOfFeatures = InvStdDev (features) : [115 x *] -> [115]
Validating --> MVNormalizedFeatures = PerDimMeanVarNormalization (features, MeanOfFeatures, InvStdOfFeatures) : [115 x *], [115], [115] -> [115 x *]
Validating --> W0*features = Times (W0, MVNormalizedFeatures) : [80 x 115], [115 x *] -> [80 x *]
Validating --> B0 = LearnableParameter() :  -> [80 x 1]
Validating --> W0*features+B0 = Plus (W0*features, B0) : [80 x *], [80 x 1] -> [80 x 1 x *]
Validating --> H1 = Tanh (W0*features+B0) : [80 x 1 x *] -> [80 x 1 x *]
Validating --> W1*H1 = Times (W1, H1) : [80 x 80], [80 x 1 x *] -> [80 x 1 x *]
Validating --> B1 = LearnableParameter() :  -> [80 x 1]
Validating --> W1*H1+B1 = Plus (W1*H1, B1) : [80 x 1 x *], [80 x 1] -> [80 x 1 x *]
Validating --> H2 = Tanh (W1*H1+B1) : [80 x 1 x *] -> [80 x 1 x *]
Validating --> W2*H2 = Times (W2, H2) : [80 x 80], [80 x 1 x *] -> [80 x 1 x *]
Validating --> B2 = LearnableParameter() :  -> [80 x 1]
Validating --> W2*H2+B2 = Plus (W2*H2, B2) : [80 x 1 x *], [80 x 1] -> [80 x 1 x *]
Validating --> H3 = Tanh (W2*H2+B2) : [80 x 1 x *] -> [80 x 
1 x *]
Validating --> W3*H2 = Times (W3, H3) : [4 x 80], [80 x 1 x *] -> [4 x 1 x *]
Validating --> B3 = L
earnableParameter() :  -> [4 x 1]
Validating --> HLast = Plus (W3*H2, B3) : [4 x 1 x *], [4 x 1] -> [4 x 1 x *]
Validating --> CrossEntropyWithSoftmax = CrossEntropyWithSoftmax (labels, HLast) : [4 x *], [4 x 1 x *] -> [1]
Validating --> EvalErrorPrediction = ErrorPrediction (labels, HLast) : [4 x *], [4 x 1 x *] -> [1]
Validating --> PosteriorProb = Softmax (HLast) : [4 x 1 x *] -> [4 x 1 x *]
Validating --> Prior = Mean (labels) : [4 x *] -> [4]
Validating --> LogOfPrior = Log (Prior) : [4] -> [4]
Validating --> ScaledLogLikelihood = Minus (HLast, LogOfPrior) : [4 x 1 x *], [4] -> [4 x 1 x *]
Validating network. 20 nodes to process in pass 2.
Validating network, final pass.
14 out of 30 nodes do not share the minibatch layout with the input data.
Post-processing network complete.
Created model with 30 nodes on GPU 0.
Training criterion node(s):
    CrossEntropyWithSoftmax = CrossEntropyWithSoftmax
Evaluation criterion node(s):
    EvalErrorPrediction = ErrorPrediction
Allocating matrices for forward and/or backward propagation.
Memory Sharing Structure:
0000000000000000: {[EvalErrorPrediction Gradient[1]] [InvStdOfFeatures Gradient[115]] [LogOfPrior Gradient[4]] [MVNormalizedFeatures Gradient[115 x *]] [MeanOfFeatures 
Gradient[115]] [PosteriorProb Gradient[4 x 1 x *]] [PosteriorProb Value[4 x 1 x *]] [Prior Gradient[4]] [ScaledLogLikelihood Gradient[4 x 1 x *]] [features Gradient[115 
x *]] [labels Gradient[4 x *]] }
000000000621A2C0: {[features Value[115 x *]] }
00000000236B9890: {[MeanOfFeatures Value[115]] }
00000000236B99D0: {[InvStdOfFeatures Value[115]] }
00000000236B9D90: {[W0 Value[80 x 115]] }
00000000236BA5B0: {[B0 Value[80 x 1]] }
00000000268876B0: {[W1 Value[80 x 80]] }
00000000268877F0: {[B1 Value[80 x 1]] }
0000000026887B10: {[W2 Value[80 x 80]] }
0000000026887C50: {[B2 Value[80 x 1]] }
0000000026887F70: {[W3 Value[4 x 80]] }
00000000268880B0: {[B3 Value[4 x 1]] }
0000000026888330: {[labels Value[4 x *]] }
00000000268885B0: {[Prior Value[4]] }
0000000026888D30: {[EvalErrorPrediction Value[1]] }
0000000026888DD0: {[ScaledLogLikelihood Value[4 x 1 x *]] }
0000000026888E70: {[CrossEntropyWithSoftmax Value[1]] }
0000000026888F10: {[LogOfPrior Value[4]] }
0000000026888FB0: {[MVNormalizedFeatures Value[115 x *]] }
0000000026889050: {[W0*features Value[80 x *]] }
00000000268890F0: {[W0 Gradient[80 x 115]] [W0*features+B0 Value[80 x 1 x *]] }
0000000026889190: {[H1 Value[80 x 1 x *]] [W0*features Gradient[80 x *]] }
0000000026889230: {[W0*features+B0 Gradient[80 x 1 x *]] [W1*H1 Value[80 x 1 x *]] }
00000000268892D0: {[W1 Gradient[80 x 80]] [W1*H1+B1 Value[80 x 1 x *]] }
0000000026889370: {[H2 Value[80 x 1 x *]] [W1*H1 Gradient[80 x 1 x *]] }
0000000026D4BF50: {[B0 Gradient[80 x 1]] [H1 Gradient[80 x 1 x *]] [W1*H1+B1 Gradient[80 x 1 x *]] [W2*H2 Value[80 x 1 x *]] }
0000000026D4BFF0: {[W2 Gradient[80 x 80]] [W2*H2+B2 Value[80 x 1 x *]] }
0000000026D4C090: {[H3 Value[80 x 1 x *]] [W2*H2 Gradient[80 x 1 x *]] }
0000000026D4C130: {[B1 Gradient[80 x 1]] [H2 Gradient[80 x 1 x *]] [W2*H2+B2 Gradient[80 x 1 x *]] [W3*H2 Value[4 x 1 x *]] }
0000000026D4C1D0: {[HLast Value[4 x 1 x *]] [W3 Gradient[4 x 80]] }
0000000026D4C590: {[CrossEntropyWithSoftmax Gradient[1]] }
0000000026D4C630: {[B2 Gradient[80 x 1]] [H3 Gradient[80 x 1 x *]] [HLast Gradient[4 x 1 x *]] }
0000000026D4C6D0: {[W3*H2 Gradient[4 x 1 x *]] }
0000000026D4C770: {[B3 Gradient[4 x 1]] }
Precomputing --> 3 PreCompute nodes found.
    MeanOfFeatures = Mean()
    InvStdOfFeatures = InvStdDev()
    Prior = Mean()
starting epoch 1 at record count 0, and file position 0
already there from last epoch
[CALL STACK]
    > 
std::locale::global
    - callnewh
    - operator new
    - 
std::vector<unsigned int,std::allocator<bool> >::_Reallocate
    - 
std::vector<float,std::allocator<float> >::push_back
    - 
UCIParser<float,std::basic_string<char,std::char_traits<char>,std::allocator<char> > >::DoneWithValue
    - 
UCIParser<float,std::basic_string<char,std::char_traits<char>,std::allocator<char> > >::Parse
    - 
Microsoft::MSR::CNTK::UCIFastReader<float>::EnsureDataAvailable
    - 
Microsoft::MSR::CNTK::UCIFastReader<float>::GetMinibatchImpl
    - 
Microsoft::MSR::CNTK::UCIFastReader<float>::TryGetMinibatch
    - 
Microsoft::MSR::CNTK::DataReaderBase::GetMinibatch
    - 
Microsoft::MSR::CNTK::DataReader::GetMinibatch
    - 
Microsoft::MSR::CNTK::DataReaderHelpers::GetMinibatchIntoNetwork<float>
    - 
Microsoft::MSR::CNTK::SGD<float>::PreCompute
    - 
Microsoft::MSR::CNTK::SGD<float>::TrainOrAdaptModel
    - 
Microsoft::MSR::CNTK::SGD<float>::Train
    - 
DoTrain<Microsoft::MSR::CNTK::ConfigParameters,float>
    - 
DoCommands<float>
EXCEPTION occurred: bad allocation
=====================================================================

When i run under the same configurations with a smaller file 2/3 of its size it works well.

Two questions here:

Q1. Is there a file size size limit versus physical memory available that we need to be aware of?
Q2. If not, is there any configuration setting i am missing in order to process this file?

Many thanks,
Thanasis

**Comment by eldakms** [comment]
UCIFastReader fetches the complete file in memory. With this reader it is not possible to process files that exceed available memory. This reader is deprecated, we encourage people to use CNTKTextFormatReader instead. You can find more information about it here https://github.com/Microsoft/CNTK/wiki/CNTKTextFormat-Reader

Thank you!
